src/js/trove/parse-pyret.js

```diff
@@ -1387,6 +107,48 @@
       "raises-violates":   function(l){return RUNTIME.getField(ast, "s-op-raises-violates").app(l);},
     }
 
+    function translate(node, fileName) {
+      let constructors = {
+        "makeNode": function(con) {
+          let args = Array.prototype.slice.call(arguments, 1);
```

Comment:
I'm not a huge fan of this, because gratuitous `arguments` munging is inefficient.  I think I'd rather see `makeNode` be curried over the constructor name, e.g. 
```js
constructors = {
  make: {
    "a-dot": RUNTIME.getField(ast, "a-dot").app
    "s-provide": RUNTIME.getField(ast, "s-provide".app
    ...
  }
}
```
so that you'd call it as `constructors.make["a-dot"](......)`.  And you could eta-expand any actual singleton values, so that e.g. `constructors.make["s-normal"]()` would just call `function() { return RUNTIME.getField(ast, "s-normal"); }`

And yes, I get that it makes the `constructors` object longer.  Perhaps the right way to do this is to iterate reflectively over the provided values out of `ast` and construct the necessary `constructors` that way.  Specifically, for each name `ast["provides"]["datatypes"][*][3][*]`, you'll get a constructor.  The name is at index [0], and if ....[1] is defined then it's a normal constructor, otherwise it's a singleton.

Reply:
Your knowledge of the internal representation seems a bit out of date :-). Here's what I was able to come up with instead:

    let nodes = {};
    const astValues = RUNTIME.getField(astLib, "defined-values");
    for (let i in astValues) {
      let value = astValues[i];
      if (value["$constructor"] !== undefined && value["$name"] === i) {
        // It's a singleton.                                                    
        // If there's a better way to tell, please change the conditional.      
        let name = value["$name"];
        nodes[name] = function() { return value; }
      }
      else if (value["$constrFor"] !== undefined) {
        // It's a normal constructor                                            
        let name = value["name"];
        nodes[name] = value.app;
      }
    }

The test to check if something's a singleton is horrific. I spent a while looking through `defined-values` and `defined-types`, and that's the best I could come up with. If there's a saner way to do this we should.

Reply:
Looking directly at the compiled output of ast.arr, the description I gave above is accurate.  It doesn't give you the `.app` closure directly, but it tells you what variants are defined for each datatype, and whether they're singletons or not.  Looking at `$constructor` or `$constrFor` is not really a stable way of doing this...